llarp/messages/link_message.hpp

```diff
@@ -7,60 +7,56 @@
 
 #include <vector>
 
+namespace
+{
+  static auto link_cat = llarp::log::Cat("lokinet.link");
+}  // namespace
+
 namespace llarp
 {
-  struct ILinkSession;
+  struct AbstractLinkSession;
   struct AbstractRouter;
 
+  inline std::array<std::string_view, 7> ABS_DICT_KEYS = {
+      "a"sv, "i"sv, "n"sv, "p"sv, "r"sv, "v"sv, "z"sv};
+
+  /// abstract base class for serialized messages
+  struct AbstractSerializable
+  {
+    virtual std::string
+    bt_encode() const = 0;
```

Comment:
i would highly suggest passing in a `std::vector<char> &` as an argument and have a non virtual helper method for returning `std::string` which just passes a local in and returns it. this way we can have a single vector that has reserved capacity to minimize heap thrashing. some of these messages, especially the relay up/down messages, are going to be insanely frequent, tens of thousands per event loop tick with potentially millions per second. having to do a heap alloc that many times is going to be noticable on the smaller end vps and small devices running as clients.

Reply:
actually, using a `llarp_buffer &` would be a way too.

Reply:
bt_{dict,list}_producer just needs a `.reserve()` to address this.

Reply:
....and i did not see that there is a `bt_encode(oxenc::bt_dict_producer&) const` method. this diff looks totally different on desktop web .... ugh